presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+            case BINARY:
+                requiredStringKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.StringSubscript.class::cast)
+                        .map(Subfield.StringSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredLongKeys = null;
+                return;
+            default:
+                throw new IllegalArgumentException("Unsupported flat map key type: " + keyOrcType);
+        }
+    }
+
+    private static Optional<TupleDomainFilter> getTopLevelFilter(Map<Subfield, TupleDomainFilter> filters)
+    {
+        Map<Subfield, TupleDomainFilter> topLevelFilters = Maps.filterEntries(filters, entry -> entry.getKey().getPath().isEmpty());
+        if (topLevelFilters.isEmpty()) {
+            return Optional.empty();
+        }
+
+        checkArgument(topLevelFilters.size() == 1, "MAP column may have at most one top-level range filter");
+        TupleDomainFilter filter = Iterables.getOnlyElement(topLevelFilters.values());
+        checkArgument(filter == IS_NULL || filter == IS_NOT_NULL, "Top-level range filter on MAP column must be IS NULL or IS NOT NULL");
+        return Optional.of(filter);
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        allNulls = false;
+
+        if (outputRequired && nullsAllowed && presentStream != null) {
+            nulls = ensureCapacity(nulls, positionCount);
+        }
+
+        if (!nullsAllowed || !nonNullsAllowed) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
+        }
+        else {
+            outputPositions = positions;
+            outputPositionsReadOnly = true;
+        }
+
+        if (presentStream == null) {
+            readNoNulls(offset, positions, positionCount);
+        }
+        else if (keyCount == 0) {
+            readAllNulls(offset, positions, positionCount);
+        }
+        else {
+            readWithNulls(offset, positions, positionCount);
+        }
+
+        localMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        return outputPositionCount;
+    }
+
+    private void readNoNulls(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!nonNullsAllowed) {
+            outputPositionCount = 0;
+            return;
+        }
+
+        nestedLengths = ensureCapacity(nestedLengths, positionCount);
+        Arrays.fill(nestedLengths, 0);
+
+        int streamPosition = 0;
+
+        for (int keyIndex = 0; keyIndex < keyCount; keyIndex++) {
+            BooleanInputStream inMapStream = inMapStreams.get(keyIndex);
+
+            if (readOffset < offset) {
```

Comment:
since the same logic is in readwithNulls too can this go into read like other readers and the one with position can be inindividual readsnulls/nofilter etc

Reply:
@bhhari I removed `readNoNulls` method and folded its logic into `readNotAllNulls`.